### PR TITLE
[wxwidgets] Update and fix

### LINKED ISSUE #25437: [wxWidgets] wxDEBUG_LEVEL should not be set to 0
**Describe the bug**
After https://github.com/microsoft/vcpkg/commit/6e16931acdf85ab5476973ba00d8eda733f42858 , wxWidgets is broken for dependent projects in Release mode, unless they also add the extra wxDEBUG_LEVEL=0 definition in building.

According to https://docs.wxwidgets.org/3.1.6/group__group__funcmacro__debug.html , the purpose of wxDEBUG_LEVEL is to remove debugging symbols completely, which almost no one does, including wxWidgets' official DLL release and its example project. That's why that commit also needs to [set wxDEBUG_LEVEL manually for examples](https://github.com/microsoft/vcpkg/commit/6e16931acdf85ab5476973ba00d8eda733f42858#diff-c2ba89066671155626d46c4112a763ba06d8ec5abbc7cab8847a2b6b8c2f421fR7).
The proper way to make a release build is to set the `NDEBUG` macro, which deactivates those debugging function, but still keep those symbols and will not cause a linker error. The commit message of 6e16931acdf85ab5476973ba00d8eda733f42858 is misleading, thost debug asserts are NOT active in Release builds. Debugging symbols are there, but they are doing nothing.
This is againt guideline ["Enable existing users of the library to switch to vcpkg"](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#enable-existing-users-of-the-library-to-switch-to-vcpkg), as it sets non-default build option for the library, and requires dependent code also set that option.

**Environment**
- OS: Windows 10.0.19044.1766
- Compiler: Visual Studio 2022 v17.2.5

**To Reproduce**
Run the Github Action of [wangqr/Aegisub](https://github.com/wangqr/Aegisub/) with vcpkg after 6e16931acdf85ab5476973ba00d8eda733f42858

**Expected behavior**
Code build without error

**Failure logs**
[gist](https://gist.github.com/wangqr/0be83aeafb831caa6a68707d334e2587)

### LINKED ISSUE #25556: [wxwidgets] Build Failed with x64-mingw-dynamic on Windows 10
Package: wxwidgets[core,sound]:x64-mingw-dynamic -> 3.1.6#2
**Host Environment**
- Host: x64-windows
- Compiler: GNU 10.3.0
-    vcpkg-tool version: 2022-06-17-9268e366206712e38102b28dbd1617697a99ff2e
    vcpkg-scripts version: 4826ede84 2022-07-01 (2 days ago)

**To Reproduce**
`vcpkg install wxwidgets`

**Failure logs**
```
-- Using cached wxWidgets-wxWidgets-35a6d7b15fedfdb5198bb6c28b31cda33b2c2a76.tar.gz.
-- Cleaning sources at C:/dev/vcpkg/buildtrees/wxwidgets/src/a33b2c2a76-d7830df458.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/dev/vcpkg/downloads/wxWidgets-wxWidgets-35a6d7b15fedfdb5198bb6c28b31cda33b2c2a76.tar.gz
-- Applying patch install-layout.patch
-- Applying patch relocatable-wx-config.patch
-- Applying patch nanosvg-ext-depend.patch
-- Applying patch fix-libs-export.patch
-- Applying patch fix-pcre2.patch
-- Using source at C:/dev/vcpkg/buildtrees/wxwidgets/src/a33b2c2a76-d7830df458.clean
-- Using cached msys-mingw-w64-i686-pkg-config-0.29.2-3-any.pkg.tar.zst.
-- Using cached msys-mingw-w64-i686-libwinpthread-git-9.0.0.6373.5be8fcd83-1-any.pkg.tar.zst.
-- Using msys root at C:/dev/vcpkg/downloads/tools/msys2/9a1ec3f33446b195
-- Found external ninja('1.10.2').
-- Configuring x64-mingw-dynamic
-- Building x64-mingw-dynamic-dbg
-- Building x64-mingw-dynamic-rel
CMake Error at ports/wxwidgets/portfile.cmake:111 (file):
  file RENAME failed to rename

    C:/dev/vcpkg/packages/wxwidgets_x64-mingw-dynamic/bin/wx-config

  to

    C:/dev/vcpkg/packages/wxwidgets_x64-mingw-dynamic/tools/wxwidgets/wx-config

  because: �t�Χ䤣����w���ɮסC

Call Stack (most recent call first):
  scripts/ports.cmake:146 (include)

```

**Additional context**
I have tried `git pull` and `vcpkg upgrade` in the vcpkg root directory, but I still can not build wxWidgets completely.  I have tried `vcpkg install wxwidget --triplet=x64-windows` and it works, but I want use the `x64-mingw-dynamic` triplet.

And I don't know why my error message looks like in the wrong encoding, I've tried `chcp 65001` which didn't work. I have confirmed that message is "No such file or directory". 

I've checked the folder `C:/dev/vcpkg/packages/wxwidgets_x64-mingw-dynamic/bin/wx-config` which didn't exist, and the folder `C:/dev/vcpkg/packages/wxwidgets_x64-mingw-dynamic/tools/wxwidgets/wx-config` is empty.

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?
  Updates wxwidgets to 3.1.7.
  Fixes #25556 (insufficient install layout patch for 3.1.6).
  Fixes #25437 (~just reverting the patch~ using a default feature for debug support).

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  unchanged, no.
  Locally tested with `vcpkg-ci-wxwidgets` for x64-linux and x64-mingw-dynamic.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  yes

### CODE REVIEW

**Comment by AenBleidd** [comment]
I don't agree with this PR. Please read my comment: https://github.com/microsoft/vcpkg/issues/25437#issuecomment-1166562207

**Comment by dg0yt** [comment]
> I don't agree with this PR. Please read my comment: [#25437 (comment)](https://github.com/microsoft/vcpkg/issues/25437#issuecomment-1166562207)

I assume you don't refer to the PR in general, but to the `wxDEBUG_LEVEL` aspect. For this aspect, my PR simply reverts the default (upstream) wxwidgets configuration as it is an immediate solution which unbreaks user projects. This is what [you suggest](https://github.com/microsoft/vcpkg/issues/25437#issuecomment-1166562207):

> it was decided that we should revert my change

This PR doesn't mean to provide a more advanced long-term solution "to control debug asserts in a more flexible way". 
Which might be the outcome of https://github.com/microsoft/vcpkg/pull/25551, if done right.

**Comment by AenBleidd** [comment]
@dg0yt, so maybe this PR and #25551 should not be merged separately then? Otherwise this PR will create regression if merged first.

File: ports/wxcharts/portfile.cmake [hunk 1]
```diff
@@ -7,8 +7,6 @@ vcpkg_from_github(
 
 vcpkg_cmake_configure(
     SOURCE_PATH "${SOURCE_PATH}"
-    OPTIONS_RELEASE
-        -DwxBUILD_DEBUG_LEVEL=0
```
Comment (JackBoosY) [inline ports/wxcharts/portfile.cmake:11]: I think we should not remove this feature since it's related to the compiler debug level.
Can you explain why?

Comment (AenBleidd) [reply ports/wxcharts/portfile.cmake:11]: @JackBoosY, please read the bug descriptions here: https://github.com/microsoft/vcpkg/issues/25437

**Comment by dg0yt** [comment]
> Otherwise this PR will create regression if merged first.

I don't see how this PR creates a regression. It adds symbols, but it doesn't limit downstream usage. This is different from #25240 which removed things, causing regressions downstream. That's why this PR is a fix, even if it some might dislike the wxwidget's default debug level.

**Comment by AenBleidd** [comment]
@dg0yt, it will lead to this error again: https://github.com/BOINC/boinc/issues/4806
When application requests complete removal of debug asserts, they will be still in wxwidgets and still enabled. Please see our discussion in the ticket I mentioned

**Comment by dg0yt** [comment]
> it will lead to this error again: [BOINC/boinc#4806](https://github.com/BOINC/boinc/issues/4806) When application requests complete removal of debug asserts, they will be still in wxwidgets and still enabled. Please see our discussion in the ticket I mentioned

Possibly yes, but this may just indicate that you didn't find the right solution.
AFAIU boinc aready sets NDEBUG, but I don't see that it ensures calling `wxDISABLE_DEBUG_SUPPORT` for release builds.
(No search hits on GH for `wxDISABLE_DEBUG_SUPPORT` or `wxIMPLEMENT_APP`, but this is what the docs says:
https://docs.wxwidgets.org/trunk/overview_debugging.html#overview_debugging_config)

And `include/wx/debug.h` says:
~~~
0:  No assertion macros at all, this should only be used when optimizing
        for resource-constrained systems (typically embedded ones)
~~~
https://github.com/wxWidgets/wxWidgets/blob/7ad1bffa875f7a38db58b6069ea3ff0c49c211d2/include/wx/debug.h#L32-L33

File: ports/wxcharts/portfile.cmake (hunk 1 above)
Comment (dg0yt) [reply ports/wxcharts/portfile.cmake:11]: It is against how wxwidgets works.

Comment (dg0yt) [reply ports/wxcharts/portfile.cmake:11]: I won't discuss or defend upstream's decisions. You can read in the docs, and in a 2009 blog post. wxwidgets wants debug level 1 as default. And consuming projects should be able to rely on the docs.

We can only discuss how to opt-out of upstream's default.
For example, if opting out is to apply for both debug and release.

**Comment by dg0yt** [comment]
For native mingw builds (on windows), I switched to the MSVC layout, so that the find module can operatore in win32 mode, not needing `sh` for `wx-config`. This reduces the amount of patching and wrapping.
Downside: there is no `wx-config` for mingw on windows. Mingw on linux is different.

File: ports/wxcharts/portfile.cmake (hunk 1 above)
Comment (AenBleidd) [reply ports/wxcharts/portfile.cmake:11]: I didn't test this but looks this proposal looks ok to me. Thanks

**Comment by JackBoosY** [comment]
Ready for review?

**Comment by dg0yt** [comment]
Configuration without debug support now testable with:
~~~
./vcpkg install --overlay-ports scripts/test_ports wxwidgets[core] vcpkg-ci-wxwidgets

**Comment by talregev** [comment]
@JackBoosY Can you review it?

**Comment by marekr** [comment]
wx 3.2 is now available fyi, basically 3.1.7 with a minor bug fixes and documentation cleanup.

There's also now a cmake config file generated rather than needing findwxwidgets

**Comment by talregev** [comment]
I think we need to do 3.1.7 then go to 3.2.0.

**Comment by talregev** [comment]
For 3.2.0 I think they will follow semver:
> wxWidgets 3.2.0 is the first release in the 3.2 stable branch. This means that it will remain ABI-compatible with all the 3.2.x releases in the future, i.e. the applications using it will continue to work with the future releases in this series even without recompiling.

Also if it almost the same, then we can do 2 PRs. one for 3.1.7 then go to 3.2.0. Then users for vcpkg can choose versions in manifests mode. More versions to choose from, in a price of 2 commit and a couple of days delay.

**Comment by dg0yt** [comment]
This PR is 3.1.7, and it is ready. Merge as is unless there is a real issue, and do 3.2.0 after that.

**Review by JackBoosY — APPROVED** [review]

**Pull request merged**